xdsl/interpreters/eqsat_pdl.py

```diff
@@ -20,3 +29,216 @@ def match_operand(
         arg = owner.operands[0]
         res = super().match_operand(ssa_val, pdl_op, arg)
         return res
+
+
+def _get_root_op(op: Operation | None) -> Operation | None:
+    """
+    Recursively finds and returns the root operation associated with the given operation.
+    """
+    return op if op is None or op.parent_op() is None else get_root_op(op.parent_op())
+
+
+@dataclass
+class EqsatPDLRewritePattern(RewritePattern):
+    functions: EqsatPDLRewriteFunctions
+    pdl_rewrite_op: pdl.RewriteOp
+    interpreter: Interpreter
+
+    def __init__(
+        self, pdl_rewrite_op: pdl.RewriteOp, ctx: MLContext, file: IO[str] | None = None
+    ):
+        pdl_pattern = pdl_rewrite_op.parent_op()
+        assert isinstance(pdl_pattern, pdl.PatternOp)
+        pdl_module = pdl_pattern.parent_op()
+        assert isinstance(pdl_module, ModuleOp)
+        self.functions = EqsatPDLRewriteFunctions(ctx)
+        self.interpreter = Interpreter(pdl_module, file=file)
+        self.interpreter.register_implementations(self.functions)
+        self.pdl_rewrite_op = pdl_rewrite_op
+
+    def match_and_rewrite(self, xdsl_op: Operation, rewriter: PatternRewriter) -> None:
+        if not self.functions.did_populate:
+            self.functions.populate_maps(cast(ModuleOp, _get_root_op(xdsl_op)))
+
+        pdl_op_val = self.pdl_rewrite_op.root
+        assert pdl_op_val is not None, "TODO: handle None root op in pdl.RewriteOp"
+        assert (
+            self.pdl_rewrite_op.body is not None
+        ), "TODO: handle None body op in pdl.RewriteOp"
```

Comment:
could you please raise DiagnosticExceptions here and add filecheck tests for them?